### PR TITLE
Implement ConnectionReuse correctly

### PR DESCRIPTION
**Pull request by tomaka**
The current implementation is a dummy. This PR replaces it with a better one.

### CODE REVIEW

File: libp2p-swarm/src/connection_reuse.rs [hunk 1]
```diff
@@ -66,31 +62,49 @@ pub struct ConnectionReuse<T, C>
...
 }
 
-struct Shared<O> {
-	// List of futures to dialed connections.
-	incoming: Vec<Box<Stream<Item = (O, Multiaddr), Error = future::SharedError<Mutex<Option<IoError>>>>>>,
-	// Tasks to signal when an element is added to `incoming`. Only used when `incoming` is empty.
-	to_signal: Vec<task::Task>,
+struct Shared<M> where M: StreamMuxer {
+	// List of active muxers.
+	active_connections: FnvHashMap<Multiaddr, M>,
```
Comment (debris) [inline libp2p-swarm/src/connection_reuse.rs]: `fnv` is faster for small keys. do you know what is the threshold? when the key is considered to be 'short' and when it is not 'short' any more? I'm just curious

File: libp2p-swarm/src/connection_reuse.rs [hunk 2]
```diff
@@ -127,31 +142,36 @@ where
 	}
 
 	fn dial(self, addr: Multiaddr) -> Result<Self::Dial, (Self, Multiaddr)> {
```
Comment (debris) [inline libp2p-swarm/src/connection_reuse.rs]: what does `Err` variant represent?  and in what cases is it returned?

File: libp2p-swarm/src/connection_reuse.rs [hunk 3]
```diff
@@ -202,125 +220,119 @@ where S: Stream<Item = (F, Multiaddr), Error = IoError>,
...
+	shared: Arc<Mutex<Shared<M>>>,
 }
 
-impl<O> Future for ConnectionReuseIncoming<O>
-	where O: Clone
+impl<M> Future for ConnectionReuseIncoming<M>
+	where M: Clone + StreamMuxer,
 {
-	type Item = (O, Multiaddr);
+	type Item = (future::FutureResult<M::Substream, IoError>, Multiaddr);
```
Comment (debris) [inline libp2p-swarm/src/connection_reuse.rs]: what does item represent and how should each case be handled?

File: libp2p-swarm/src/swarm.rs [hunk 4]
```diff
@@ -233,7 +235,7 @@ impl<T, C, H, If, F> Future for SwarmFuture<T, C, H, F>
                     self.listeners.push(listener);
                 },
                 Ok(Async::Ready(None)) => {},
-                Err(err) => return Err(err),
+                Err(_err) => {},        // Ignoring errors
```
Comment (debris) [inline libp2p-swarm/src/swarm.rs]: why are errors being ignored here?

File: libp2p-swarm/src/connection_reuse.rs [hunk 5]
```diff
@@ -202,125 +220,119 @@ where S: Stream<Item = (F, Multiaddr), Error = IoError>,
...
-		let mut connections_to_drop = upgrades_to_drop;
-
-		for (index, &mut (ref mut muxer, ref mut next_incoming, ref client_addr)) in
-			self.connections.iter_mut().enumerate()
-		{
+		// Check whether any incoming substream is ready.
+		// We extract everything at the start, then insert back the elements that we still want at
+		// the next iteration.
+        for n in (0 .. self.connections.len()).rev() {
+            let (muxer, mut next_incoming, client_addr) = self.connections.swap_remove(n);
```
Comment (debris) [inline libp2p-swarm/src/connection_reuse.rs]: formatting is inconsistent here (spaces)

File: libp2p-swarm/src/connection_reuse.rs (hunk 1 above)
Comment (tomaka) [reply libp2p-swarm/src/connection_reuse.rs]: I found this (old) link: http://cglab.ca/~abeinges/blah/hash-rs/
Most the time the multiaddr will be something like `/ip4/a.b.c.d/tcp/e` which is 8 bytes. Fnv may indeed be more appropriate here.

File: libp2p-swarm/src/swarm.rs (hunk 4 above)
Comment (tomaka) [reply libp2p-swarm/src/swarm.rs]: An error happens if the remote sends us bad data, so we just drop the connection. This probably needs a comment and a log entry.

File: libp2p-swarm/src/connection_reuse.rs (hunk 2 above)
Comment (tomaka) [reply libp2p-swarm/src/connection_reuse.rs]: Happens if the multiaddress is invalid or not supported. This should be documented in the `Transport` trait if it isn't the case.

**Comment by tomaka** [comment]
Closing in favour of #117

**Pull request closed without merging**